### PR TITLE
[Provisioner] Support multi level performance disk

### PR DESCRIPTION
**Pull request by cblmemo**
This PR introduces multi-level performance disks in all clouds. By specifying `disk_tier` in `sky.Resources` to one of `high`, `medium`, and `low`, a user could use a disk with custom performance. All clouds' current decisions and prices (256 GB for example) are shown below.

`disk_tier=low`
|Cloud|Disk Type|Benchmarked Read Throughput|Benchmarked Read IOPS|Benchmarked Write Throughput|Benchmarked Write IOPS|Price (GiB/mo)|Price (IOPS/mo)|Total Price Each Month|
|-|-|-|-|-|-|-|-|-|
|GCP|pd-standard|13.87 MB/s|211.63|37.17 MB/s|567.13|0.048|-|12.288|
|Azure|Standard_LRS|861.92 MB/s|13151.92|36.67 MB/s|559.54|-|-|11.328|
|AWS|standard|20.33 MB/s|310.14|282.19 MB/s|4305.91|0.066|-|16.896|

`disk_tier=medium`
|Cloud|Disk Type|Benchmarked Read Throughput|Benchmarked Read IOPS|Benchmarked Write Throughput|Benchmarked Write IOPS|Price (GiB/mo)|Price (IOPS/mo)|Total Price Each Month|
|-|-|-|-|-|-|-|-|-|
|GCP|pd-balanced|222.86 MB/s|3400.58|222.25 MB/s|3391.25|0.12|-|30.72|
|Azure|Premium_LRS|211.80 MB/s|3231.88|176.12 MB/s|2687.44|-|-|36.29|
|AWS|gp3 with IOPS=3500|241.43 MB/s|3683.87|199.15 MB/s|3038.86|0.09|0.005|25.54|

`disk_tier=high`
|Cloud|Disk Type|Benchmarked Read Throughput|Benchmarked Read IOPS|Benchmarked Write Throughput|Benchmarked Write IOPS|Price (GiB/mo)|Price (IOPS/mo)|Total Price Each Month|
|-|-|-|-|-|-|-|-|-|
|GCP|pd-ssd|342.91 MB/s|5232.42|268.17 MB/s|4091.91|0.203|-|51.968|
|Azure|-|-|-|-|-|-|-|-|
|AWS|gp3 with IOPS=7000|323.54 MB/s|4936.80|254.82 MB/s|3888.22|0.09|0.005|43.04|

Azure `disk_tier=high` is disabled now since a high-performance disk in Azure cannot be launched as an OS disk.

Benchmark command:

```shell
fio --name=sync_rand_64k_{read, write}s  --rw=rand{read, write}  --direct=1 --ioengine=sync --bs=64k --numjobs=4 --iodepth=128 --size=1G --group_reporting --directory=/tmp/
```

Tested (run the relevant ones):

- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by Michaelvll** [comment]
Awesome! Thanks for submitting the PR. Will get to it soon. Could you also add the performance (claimed by cloud, and benchmarked result) for each disk type to the table in your PR description, so we can have an idea of how close each tier would be? Thanks!

File: sky/resources.py
Comment (Michaelvll) [inline sky/resources.py:309]: If we move the configuration into `cloud.make_deploy_resources_variables()` as mentioned above, so we don't have to do this here.

File: sky/resources.py
Comment (Michaelvll) [inline sky/resources.py:996]:
```suggestion
        if version < 9:
            self._disk_type = None
```

Also, increase the `Resources.Version`

File: sky/skylet/providers/azure/azure-vm-template.json
Comment (Michaelvll) [inline sky/skylet/providers/azure/azure-vm-template.json:95]: Is it able to make this choices of the three types, instead of the general string?

File: sky/skylet/providers/azure/azure-vm-template.json
Comment (Michaelvll) [inline sky/skylet/providers/azure/azure-vm-template.json:209]: Nice! Great work for figuring this out!

File: sky/templates/aws-ray.yml.j2
Comment (Michaelvll) [inline sky/templates/aws-ray.yml.j2:53]: Why do we remove this throughput? It seems the AWS allows to specify the Iops and Throughput separately. Should we consider both in our disk type separation?

File: sky/templates/gcp-ray.yml.j2
Comment (Michaelvll) [inline sky/templates/gcp-ray.yml.j2:57]: Need to confirm: It seems that I cannot stop the instance when the `disk_type` is specified as `pd-ssd`. Could you confirm if that is true? If that is the case we may need to think about how to get around that.

**Review by Michaelvll — COMMENTED** [review]
Awesome @cblmemo! Thanks for adding this in a short time! Left some initial comments. : )

**Comment by cblmemo** [comment]
> Awesome! Thanks for submitting the PR. Will get to it soon. Could you also add the performance (claimed by cloud, and benchmarked result) for each disk type to the table in your PR description, so we can have an idea of how close each tier would be? Thanks!

🫡 will upload results asap

File: sky/resources.py
Comment (cblmemo) [reply sky/resources.py:300]: I'm wondering whether it is feasible to add `specify_disk_type` in `requested_features` since other features are all task-level features and this is a resource-level feature. That is to say, multi resources in the same task may not require this feature simultaneously. 🤔
Currently, I move this check into `_try_validate_disk_type`.

File: sky/templates/aws-ray.yml.j2
Comment (cblmemo) [reply sky/templates/aws-ray.yml.j2:53]: When I try provision for AWS, I found that some disk type does not support specifying throughput. Also, throughput will be auto-specified from IOPS, so I remove it.

File: sky/clouds/aws.py
Comment (Michaelvll) [inline sky/clouds/aws.py:324]: These functions seem only used inside this module, we can add the prefix `_`  to the function name to make them private. reference: https://google.github.io/styleguide/pyguide.html#3162-naming-conventions

File: sky/clouds/aws.py
Comment (Michaelvll) [inline sky/clouds/aws.py:588]: Can we make these functions a single one `_get_disk_specs`, as they are only used once and we may want to avoid shallow functions?
```
def _get_disk_specs(...) -> Dict[str, Any]:
    ...
```

File: sky/clouds/aws.py
Comment (Michaelvll) [reply sky/clouds/aws.py:324]: Actually, as mentioned below we can changes these separate calls to:

```suggestion
            **AWS._get_disk_specs(...)
```

File: sky/backends/backend_utils.py
Comment (Michaelvll) [inline sky/backends/backend_utils.py:841]: Correct me if I am wrong, we don't have to prompt the return value up to here. Instead, having the disk type in the `resources_vars` in L836 should be enough, as the upper level logics may not need to know the cloud specific disk_type specs.

File: sky/clouds/cloud.py
Comment (Michaelvll) [inline sky/clouds/cloud.py:455]: Don't have to expose this.

File: sky/clouds/azure.py
Comment (Michaelvll) [inline sky/clouds/azure.py:444]: Don't have to expose this.

File: sky/backends/backend_utils.py
Comment (Michaelvll) [reply sky/backends/backend_utils.py:841]: One principle for the class design is to reduce the number of interfaces. : )

File: sky/resources.py
Comment (Michaelvll) [inline sky/resources.py:171]: We should hide the actual lower-level descriptions from the users for a better user experience.

File: sky/resources.py
Comment (Michaelvll) [inline sky/resources.py:287]: Showing the disk_type specifed in the yaml `low`, `medium`, `high` is good enough. We should not show the details.

File: sky/templates/aws-ray.yml.j2
Comment (Michaelvll) [reply sky/templates/aws-ray.yml.j2:53]: Can we explain why the throughput is added back again?

File: sky/clouds/cloud.py
Comment (Michaelvll) [inline sky/clouds/cloud.py:465]: Let's avoid showing this detailed information for now. Instead, we may want to explain the disk_type spec in our doc with the brief note about the performance: https://github.com/skypilot-org/skypilot/blob/master/docs/source/reference/yaml-spec.rst

File: sky/resources.py
Comment (Michaelvll) [inline sky/resources.py:1]: We should also fix the `less_demanding_than` function, so that a job requires `high` disk type will not be able to run on the cluster with `low` disk type.

**Review by Michaelvll — COMMENTED** [review]
Thanks for fixing the PR @cblmemo! It looks pretty close to being ready. Left several comments mostly for better code style.
Several general comments:
1. If we try to launch a VM with `disk_type: high`, will it successfully failover through all the clouds, instead of directly failing when we meet Azure or Lambda? We can actually try it by `sky launch --gpus A100:8` to see if it will failover through the clouds.
2. Can we add a CLI as well? We should have a `sky launch --disk-type` option.

File: sky/templates/aws-ray.yml.j2
Comment (cblmemo) [reply sky/templates/aws-ray.yml.j2:53]: Because now I only enable it when disk_type == gp3. Please refer to the `custom_disk_perf` flag.

**Comment by concretevitamin** [comment]
Skimming the description, this looks great! One quick thing I noticed is `disk_type=low` maps to very different performance on the three clouds. Is this desirable or by design? 

Asking because previous users have found that the default GCP disk is too slow. If we default to `medium`, this may be fine.

**Comment by Michaelvll** [comment]
> Skimming the description, this looks great! One quick thing I noticed is `disk_type=low` maps to very different performance on the three clouds. Is this desirable or by design?

Please correct me if I am wrong @cblmemo. That is the best compromisation we can get to match the performance among all three clouds, as the worst disks on those clouds have very different performances.
It should be fine, as we just guarantee the lower bound of the disk performance.

> Asking because previous users have found that the default GCP disk is too slow. If we default to `medium`, this may be fine.

Yep, good idea. Defaulting to medium is a good choice.

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:229]:
```suggestion
                                   '"medium", "high". Default: medium'))
```

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:1188]:
```suggestion
              help=('OS disk type. Could be one of "low", "medium", "high". Default: medium'))
```

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:3257]:
```suggestion
              help=('OS disk type. Could be one of "low", "medium", "high". Default: medium'))
```

File: sky/cli.py
Comment (Michaelvll) [inline sky/cli.py:3608]:
```suggestion
              help=('OS disk type. Could be one of "low", "medium", "high". Default: medium'))
```

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix @cblmemo! The latest code looks good to me. I am testing it for now. 

Several questions: 
1. should we update `sky/utils/schemas.py` to ensure the `resources` field can take the `disk_tier` argument?
2. If we have the current PR merged, will #1708, #1744 need to be updated?

File: sky/clouds/aws.py
Comment (cblmemo) [reply sky/clouds/aws.py]: `iops` and `throughput` won't appear in aws-ray.yml after jinja2 rendering. I think the unexpected `iops` arguments is a bug of boto3 and the bug won't appear if we specify `standard` as volume type

**Comment by cblmemo** [comment]
> Thanks for the quick fix @cblmemo! The latest code looks good to me. I am testing it for now.
> 
> Several questions:
> 
> 1. should we update `sky/utils/schemas.py` to ensure the `resources` field can take the `disk_tier` argument?
> 2. If we have the current PR merged, will [[Optimizer] Add GCP disk price to the optimizer #1708](https://github.com/skypilot-org/skypilot/pull/1708), [[Optimizer] Add Azure disk price to the optimizer #1744](https://github.com/skypilot-org/skypilot/pull/1744) need to be updated?

For the second question, yes, we need to add multi-tier prices into the disk price calculation. The previous implementation just fetches the default tier disk type's price and uses it to calculate the total price.

File: sky/clouds/aws.py
Comment (Michaelvll) [inline sky/clouds/aws.py:320]: It might be better to have a constant called `_DEFAULT_DISK_TIER` in the `Cloud` class, and let the inherited classes use that.
Also, it might be better to handle the None->'_DEFAULT_DISK_TIER` inside the method `_get_disk_specs` to reduce the abstraction leakage. : )

File: sky/clouds/azure.py
Comment (Michaelvll) [inline sky/clouds/azure.py:241]: ditto

File: sky/clouds/gcp.py
Comment (Michaelvll) [inline sky/clouds/gcp.py:355]: ditto

**Review by Michaelvll — COMMENTED** [review]
Thanks for the quick fix! The code looks pretty good with a small nit.
A problem I met during testing:
1. `sky launch --disk-tier medium`. This will error out due to the Azure `instance_type` not supporting the disk-tier, which is unexpected, as we should let Azure create CPU instance. Also, it seems inconsistent that if I use `sky launch` directly, there will be no error. Is it safe to let Azure default to medium for CPU instances?

**Review by Michaelvll — APPROVED** [review]
Thank you for adding this fantastic feature @cblmemo! This will significantly improve the performance of large model training.

Several TODOs (maybe in the future PR, please file issues for them if we decide to leave it to the future):
- [ ] Identity the performance of the lambda's disk performance and enable it when the user specifies the corresponding disk_tier.
- [ ] Add disk cost into the optimizer

File: sky/clouds/aws.py
Comment (cblmemo) [reply sky/clouds/aws.py:556]: It is some interim code that I forget to change back. Fixed now 🫡

File: sky/clouds/azure.py
Comment (cblmemo) [reply sky/clouds/azure.py:482]: It is used in the smoke test

File: tests/test_smoke.py
Comment (cblmemo) [reply tests/test_smoke.py:1661]: I see. Fixed now

File: sky/clouds/azure.py
Comment (cblmemo) [reply sky/clouds/azure.py:452]: Oh yes... Let me find a more precise algorithm and test for all instance types

**Review by Michaelvll — APPROVED** [review]
Thanks for the quick fix @cblmemo! It looks good to me. Let's merge this if the master branch is merged and smoke tests are all passed.

**Comment by cblmemo** [comment]
I've resolved the merge conflict and passed all smoke tests. I think it is ready to merge. 🫡

**Comment by Michaelvll** [comment]
Thanks for the awesome work and the quick fixes @cblmemo! The PR looks good to me and I also tested it with the smoke tests on GCP, manually tried several combinations of the options with the disk tier, and launched a VM on azure. They all look good! Merging the PR now.

**Pull request merged**
